### PR TITLE
[gem] Expose sidekiq metrics to prometheus

### PR DESCRIPTION
**Pull request by hallelujah**
Closes https://github.com/3scale/infrastructure/issues/574

The metrics are exposed as:

http://${MASTER_DOMAIN}/system/metrics

Metrics are only available if authentication basic is configured by setting in the `config/prometheus.yml` file the `username` and `password`

### CODE REVIEW

**Comment by maneta** [comment]
@hallelujah AFAIK Prometheus does not support auth-basic for the scraping.

Ideally, the option would be to run the exporter in a different port without authentication and then we would be able to secure this port by security group and in case of Openshift we could secure it at a service or pod levels and never expose the route.

**Comment by slopezz** [comment]
AFAIK, prometheus does not support even Auth basic, and you need to use a reverse proxy to do so:

https://prometheus.io/docs/guides/basic-auth/

What normally is done is to expose this metrics on a different port (like 9292, each exporter uses a different port), and protect that access using security groups for example.

**Comment by maneta** [comment]
@slopezz yes we could use a nginx in SAAS but for the on-prem product that would be an ugly boilerplate and in both cases a bottleneck.

**Comment by hallelujah** [comment]
OK now I got what is Prometheus (thanks for the call guys)

* I need to listen to another port on unicorn
* Using the default port for sidekiq for now as mentionned in  https://github.com/prometheus/prometheus/wiki/Default-port-allocations

**Comment by mikz** [comment]
@maneta @slopezz https://prometheus.io/docs/guides/basic-auth/#testing shows prometheus can use Basic Auth for scraping just fine.
https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

Sure, it might not work with the kubernetes_sd_config, but there we can expose a separate port for metrics.

**Comment by mikz** [comment]
@hallelujah No to start listening on different port (for saas for example), over time system would expose other metrics too, so it would have just one endpoint for the prometheus metrics.

It is easy to mount it anywhere as it is just a rack app.

Yes, on-prem we would probably start it on a second port, but that can be done during some prometheus project where we align all our projects to use it.

**Comment by hallelujah** [comment]
@mikz That was the first intention but @maneta told me we should not expose metrics publicly and prometheus does not handle authentication so we were left to listening another port.

**Comment by slopezz** [comment]
On other applications like apicast, https://github.com/3scale/3scale-amp-openshift-templates/blob/master/apicast-gateway/apicast.yml#L81 ,  we open on the same container 3 different ports:
- 8080/8090 are app ports
- 9421 is the prometheus apicast exporter

**Comment by mikz** [comment]
@hallelujah as said in 
https://github.com/3scale/porta/pull/274#issuecomment-434768077 it does handle basic authentication, so for saas it is just fine.

@slopezz other apps like APIcast are not Rails. That generally expects to be listening just on one port. The same way as for example zync.

**Comment by maneta** [comment]
> @maneta @slopezz https://prometheus.io/docs/guides/basic-auth/#testing shows prometheus can use Basic Auth for scraping just fine.
> https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config
> 
> Sure, it might not work with the kubernetes_sd_config, but there we can expose a separate port for metrics.

Yes, @mikz I taught that we would need to deploy a nginx for that. But yes looks like the scrape_config can take care of the config. Good!

**Comment by maneta** [comment]
@hallelujah this is the info the documentation give us on auth basic:

```
# Sets the `Authorization` header on every scrape request with the
# configured username and password.
# password and password_file are mutually exclusive.
basic_auth:
  [ username: <string> ]
  [ password: <secret> ]
  [ password_file: <string> ]

# Sets the `Authorization` header on every scrape request with
# the configured bearer token. It is mutually exclusive with `bearer_token_file`.
[ bearer_token: <secret> ]

# Sets the `Authorization` header on every scrape request with the bearer token
# read from the configured file. It is mutually exclusive with `bearer_token`.
[ bearer_token_file: /path/to/bearer/token/file ]
```

Source: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config

**Comment by hallelujah** [comment]
@maneta @slopezz @mikz Good to review.
Few questions for polishing:

- should also check the Realm in the constraint? I am not really sure about the value
- @mikz constraints return 404 though to be correct it should return 401 but I think this is OK as the way to do that would be to have a controller or a at least rack app with that logic.

File: config/routes.rb [hunk 1]
```diff
@@ -11,7 +11,12 @@ def call(env)
   end
 end
 
+
 System::Application.routes.draw do
+  constraints MasterDomainWithAuthBasicConstraint.new(ThreeScale.config.prometheus.username, ThreeScale.config.prometheus.password) do
+    mount Sidekiq::Prometheus::Exporter.to_app, at: '/system'
```
Comment (mikz) [inline config/routes.rb]: Usually the endpoint is called `/metrics`.

Comment (mikz) [reply config/routes.rb]: But it is probably for the best to namespace it.

File: .rubocop.yml [hunk 2]
```diff
@@ -115,7 +115,7 @@ Layout/SpaceInsideHashLiteralBraces:
 
 # Offense count: 514
 # Cop supports --auto-correct.
-Style/SpaceInsideBrackets:
+Layout/SpaceInsideReferenceBrackets:
```
[the remark this answers is not in the document]
Comment (hallelujah) [reply .rubocop.yml]: Now everyone gets it

**Pull request merged**